### PR TITLE
Limit the maximum amount of ancestors in a ZIO Trace

### PR DESCRIPTION
**Pull request by kaishh**
* Add `ZTrace#parents`

This cures `StackOverflowError`s of form:
```
Exception in thread "specs2.user-2" java.lang.StackOverflowError
	at java.base/java.lang.StringBuilder.<init>(StringBuilder.java:99)
	at java.base/java.lang.StackTraceElement.toString(StackTraceElement.java:356)
	at zio.internal.stacktracer.ZTraceElement$SourceLocation.prettyPrint(ZTraceElement.scala:31)
	at zio.ZTrace.$anonfun$prettyPrint$1(ZTrace.scala:34)
	at scala.collection.immutable.List.map(List.scala:286)
	at zio.ZTrace.prettyPrint(ZTrace.scala:34)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
	at zio.ZTrace.prettyPrint(ZTrace.scala:45)
	at zio.ZTrace.$anonfun$prettyPrint$3(ZTrace.scala:46)
	at scala.Option.map(Option.scala:163)
```

I think there may still be a dangerous case with too many parallel traces, but not sure about (must reproduce separately)

### CODE REVIEW

File: core/shared/src/main/scala/zio/internal/FiberContext.scala [hunk 1]
```diff
@@ -128,10 +128,30 @@ private[zio] final class FiberContext[E, A](
   private[this] final def cutAncestryTrace(trace: ZTrace): ZTrace = {
     val maxExecLength  = platform.tracing.tracingConfig.ancestorExecutionTraceLength
     val maxStackLength = platform.tracing.tracingConfig.ancestorStackTraceLength
+    val maxAncestors   = platform.tracing.tracingConfig.ancestryLength - 1
+
+    val flatAncestors = {
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/internal/FiberContext.scala]: Can we push this out into a helper method, maybe in `ZTrace` companion object?

**Comment by jdegoes** [comment]
One failed test, probably rerunning will fix it.

File: core/shared/src/main/scala/zio/internal/FiberContext.scala (hunk 1 above)
Comment (kaishh) [reply core/shared/src/main/scala/zio/internal/FiberContext.scala]: Replaced by a call to `trace.parents`

**Comment by kaishh** [comment]
@jdegoes Ready for another review

File: core/shared/src/main/scala/zio/internal/FiberContext.scala [hunk 2]
```diff
@@ -128,10 +128,21 @@ private[zio] final class FiberContext[E, A](
   private[this] final def cutAncestryTrace(trace: ZTrace): ZTrace = {
     val maxExecLength  = platform.tracing.tracingConfig.ancestorExecutionTraceLength
     val maxStackLength = platform.tracing.tracingConfig.ancestorStackTraceLength
+    val maxAncestors   = platform.tracing.tracingConfig.ancestryLength - 1
 
-    trace.copy(
+    val cutParentTrace =
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/internal/FiberContext.scala]: This looks much better! Can we put `def truncateTrace(list: List[ZTrace], max: Int)` inside `ZTrace` companion object as well? Easier to test independently and this helps cut down on size of `FiberContext`.

Comment (kaishh) [reply core/shared/src/main/scala/zio/internal/FiberContext.scala]: Done!

**Comment by kaishh** [comment]
> [error]     x supervision inheritance
> [error]  'ArraySeq()' doesn't have size 1 but size 0 (RTSSpec.scala:996)

**Pull request merged**